daemon/nvidia_linux.go

```diff
@@ -38,70 +40,90 @@ func init() {
 		capset:     capset,
 		updateSpec: setNvidiaGPUs,
 	}
-	for c := range capset {
-		nvidiaDriver.capset[c] = struct{}{}
+	for c := range allNvidiaCaps {
+		nvidiaDriver.capset[string(c)] = struct{}{}
 	}
 	registerDeviceDriver("nvidia", nvidiaDriver)
 }
 
 func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
-	var opts []nvidia.Opts
+	envMap := getEnvMap(s.Process.Env)
 
 	req := dev.req
 	if req.Count != 0 && len(req.DeviceIDs) > 0 {
 		return errConflictCountDeviceIDs
 	}
 
 	if len(req.DeviceIDs) > 0 {
-		var ids []int
-		var uuids []string
-		for _, devID := range req.DeviceIDs {
-			id, err := strconv.Atoi(devID)
-			if err == nil {
-				ids = append(ids, id)
-				continue
-			}
-			// if not an integer, then assume UUID.
-			uuids = append(uuids, devID)
-		}
-		if len(ids) > 0 {
-			opts = append(opts, nvidia.WithDevices(ids...))
-		}
-		if len(uuids) > 0 {
-			opts = append(opts, nvidia.WithDeviceUUIDs(uuids...))
-		}
-	}
-
-	if req.Count < 0 {
-		opts = append(opts, nvidia.WithAllDevices)
+		envMap["NVIDIA_VISIBLE_DEVICES"] = strings.Join(req.DeviceIDs, ",")
 	} else if req.Count > 0 {
-		opts = append(opts, nvidia.WithDevices(countToDevices(req.Count)...))
+		envMap["NVIDIA_VISIBLE_DEVICES"] = countToDevices(req.Count)
+	} else if req.Count < 0 {
+		envMap["NVIDIA_VISIBLE_DEVICES"] = "all"
 	}
 
-	var nvidiaCaps []nvidia.Capability
+	var nvidiaCaps []string
 	// req.Capabilities contains device capabilities, some but not all are NVIDIA driver capabilities.
 	for _, c := range dev.selectedCaps {
 		nvcap := nvidia.Capability(c)
 		if _, isNvidiaCap := allNvidiaCaps[nvcap]; isNvidiaCap {
-			nvidiaCaps = append(nvidiaCaps, nvcap)
+			nvidiaCaps = append(nvidiaCaps, c)
 			continue
 		}
 		// TODO: nvidia.WithRequiredCUDAVersion
 		// for now we let the prestart hook verify cuda versions but errors are not pretty.
 	}
 
 	if nvidiaCaps != nil {
-		opts = append(opts, nvidia.WithCapabilities(nvidiaCaps...))
+		envMap["NVIDIA_DRIVER_CAPABILITIES"] = strings.Join(nvidiaCaps, ",")
+	}
+
+	s.Process.Env = envMapToArray(envMap)
+
+	path, err := exec.LookPath(nvidiaHook)
+	if err != nil {
+		return err
 	}
 
-	return nvidia.WithGPUs(opts...)(nil, nil, nil, s)
+	if s.Hooks == nil {
+		s.Hooks = &specs.Hooks{}
+	}
+	s.Hooks.Prestart = append(s.Hooks.Prestart, specs.Hook{
+		Path: path,
+		Args: []string{
+			nvidiaHook,
+			"prestart",
+		},
+		Env: os.Environ(),
+	})
+
+	return nil
 }
 
 // countToDevices returns the list 0, 1, ... count-1 of deviceIDs.
-func countToDevices(count int) []int {
-	devices := make([]int, count)
+func countToDevices(count int) string {
+	devices := make([]string, count)
 	for i := range devices {
-		devices[i] = i
+		devices[i] = strconv.Itoa(i)
+	}
+	return strings.Join(devices, ",")
+}
+
+func getEnvMap(e []string) map[string]string {
+	m := make(map[string]string)
+	for _, s := range e {
+		p := strings.SplitN(s, "=", 2)
+		m[p[0]] = p[1]
```

Comment:
This will panic if there's no `=` in `s`

Reply:
Thanks for noticing! Fixed